File: backend-sdk/topology/topology.go
package topology

import "github.com/flowcanvas/flowcanvas-backend-sdk/graph"

type Result struct {
	Order         []graph.NodeID   `json:"order"`
	Layers        [][]graph.NodeID `json:"layers"`
	CyclicNodeIDs []graph.NodeID   `json:"cyclicNodeIds"`
}

func Analyze(g *graph.GraphDocument) Result {
	if g == nil {
		return Result{}
	}
	nodeOrder := make([]graph.NodeID, 0, len(g.Nodes))
	nodeSet := map[graph.NodeID]struct{}{}
	inDegree := map[graph.NodeID]int{}
	outgoing := map[graph.NodeID][]graph.Edge{}
	for _, node := range g.Nodes {
		if _, exists := nodeSet[node.ID]; exists {
			continue
		}
		nodeOrder = append(nodeOrder, node.ID)
		nodeSet[node.ID] = struct{}{}
		inDegree[node.ID] = 0
		outgoing[node.ID] = []graph.Edge{}
	}
	for _, edge := range g.Edges {
		if _, ok := nodeSet[edge.Source]; !ok {
			continue
		}
		if _, ok := nodeSet[edge.Target]; !ok {
			continue
		}
		outgoing[edge.Source] = append(outgoing[edge.Source], edge)
		inDegree[edge.Target]++
	}

	frontier := make([]graph.NodeID, 0)
	for _, id := range nodeOrder {
		if inDegree[id] == 0 {
			frontier = append(frontier, id)
		}
	}

	order := make([]graph.NodeID, 0, len(nodeOrder))
	layers := [][]graph.NodeID{}
	processed := map[graph.NodeID]struct{}{}
	for len(frontier) > 0 {
		layer := append([]graph.NodeID(nil), frontier...)
		layers = append(layers, layer)
		next := []graph.NodeID{}
		for _, id := range frontier {
			order = append(order, id)
			processed[id] = struct{}{}
			for _, edge := range outgoing[id] {
				inDegree[edge.Target]--
				if inDegree[edge.Target] == 0 {
					next = append(next, edge.Target)
				}
			}
		}
		frontier = next
	}

	cyclic := []graph.NodeID{}
	for _, id := range nodeOrder {
		if _, ok := processed[id]; !ok {
			cyclic = append(cyclic, id)
		}
	}

	return Result{Order: order, Layers: layers, CyclicNodeIDs: cyclic}
}

func WouldCreateCycle(g *graph.GraphDocument, source, target graph.NodeID) bool {
	if source == target {
		return true
	}
	outgoing := map[graph.NodeID][]graph.NodeID{}
	for _, node := range g.Nodes {
		outgoing[node.ID] = []graph.NodeID{}
	}
	for _, edge := range g.Edges {
		outgoing[edge.Source] = append(outgoing[edge.Source], edge.Target)
	}
	outgoing[source] = append(outgoing[source], target)

	stack := []graph.NodeID{target}
	visited := map[graph.NodeID]struct{}{}
	for len(stack) > 0 {
		current := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if current == source {
			return true
		}
		if _, ok := visited[current]; ok {
			continue
		}
		visited[current] = struct{}{}
		stack = append(stack, outgoing[current]...)
	}
	return false
}

File: frontend-sdk/src/core/topology.ts
import type { CanvasEdge, GraphDocument, NodeId } from './types';

export interface TopologyResult {
  order: NodeId[];
  layers: NodeId[][];
  cyclicNodeIds: NodeId[];
}

export function analyzeTopology(graph: Pick<GraphDocument, 'nodes' | 'edges'>): TopologyResult {
  const nodeIds = new Set(graph.nodes.map(node => node.id));
  const inDegree = new Map<NodeId, number>();
  const outgoing = new Map<NodeId, CanvasEdge[]>();

  for (const id of nodeIds) {
    inDegree.set(id, 0);
    outgoing.set(id, []);
  }

  for (const edge of graph.edges) {
    if (!nodeIds.has(edge.source) || !nodeIds.has(edge.target)) continue;
    outgoing.get(edge.source)?.push(edge);
    inDegree.set(edge.target, (inDegree.get(edge.target) ?? 0) + 1);
  }

  let frontier = [...nodeIds].filter(id => inDegree.get(id) === 0);
  const order: NodeId[] = [];
  const layers: NodeId[][] = [];
  const processed = new Set<NodeId>();

  while (frontier.length) {
    // Map/Set insertion order is deterministic for a given graph, so sorting
    // every layer is unnecessary and would turn this into O(N log N).
    const layer = frontier;
    layers.push(layer);
    const next: NodeId[] = [];
    for (const id of layer) {
      order.push(id);
      processed.add(id);
      for (const edge of outgoing.get(id) ?? []) {
        const degree = (inDegree.get(edge.target) ?? 0) - 1;
        inDegree.set(edge.target, degree);
        if (degree === 0) next.push(edge.target);
      }
    }
    frontier = next;
  }

  return {
    order,
    layers,
    cyclicNodeIds: [...nodeIds].filter(id => !processed.has(id)),
  };
}

export function wouldCreateCycle(
  graph: Pick<GraphDocument, 'nodes' | 'edges'>,
  source: NodeId,
  target: NodeId,
): boolean {
  if (source === target) return true;
  const outgoing = new Map<NodeId, NodeId[]>();
  for (const node of graph.nodes) outgoing.set(node.id, []);
  for (const edge of graph.edges) outgoing.get(edge.source)?.push(edge.target);
  outgoing.get(source)?.push(target);

  const stack = [target];
  const visited = new Set<NodeId>();
  while (stack.length) {
    const current = stack.pop()!;
    if (current === source) return true;
    if (visited.has(current)) continue;
    visited.add(current);
    stack.push(...(outgoing.get(current) ?? []));
  }
  return false;
}
